### DISCUSSION TITLE
Memory leak using Octane?

### DISCUSSION THREAD

**Discussion by EdgarsJoja** [livewire/livewire#8338] [1. Help]
Hi, I'm having some issues with Livewire + Octane setup - looks like a memory leak.
I created simplified test component within my project to try and isolate the issue.

Livewire component
```php
class Test extends Component
{
    use MemoryFootprintLogger;

    public function render(): View
    {
        return view('livewire.test');
    }

    #[Computed]
    protected function getLargeArray(): array
    {
        return array_fill(0, 1000000, 'test');
    }
}
```

Trait to dump memory usage for components
```php
trait MemoryFootprintLogger
{
    public function bootMemoryFootprintLogger(): void
    {
        dump($this->memoryReport('boot'));
    }

    public function mountMemoryFootprintLogger(): void
    {
        dump($this->memoryReport('mount'));
    }

    public function renderingMemoryFootprintLogger(): void
    {
        dump($this->memoryReport('rendering'));
    }

    public function renderedMemoryFootprintLogger(): void
    {
        dump($this->memoryReport('rendered'));
    }

    protected function memoryUsageInMegaBytes(): int
    {
        return (int)(memory_get_peak_usage() / (1024 * 1024));
    }

    protected function memoryReport(string $method): string
    {
        return sprintf('%s (%s) using %s MB', $this::class, $method, $this->memoryUsageInMegaBytes());
    }
}
```

Component blade template 
```blade
<div>
    Test {{ count($this->getLargeArray) }}
</div>
```

Steps to replicate:
1. Render component;
2. Check memory usage first time (even this was high, probably due to previous tests - normally it would be `1mb` up to `rendering` and  ~`17mb` for `rendered`):
```
"App\Livewire\Test (boot) using 6 MB" // app/Livewire/Debug/MemoryFootprintLogger.php:11

"App\Livewire\Test (mount) using 6 MB" // app/Livewire/Debug/MemoryFootprintLogger.php:16

"App\Livewire\Test (rendering) using 6 MB" // app/Livewire/Debug/MemoryFootprintLogger.php:21

"App\Livewire\Test (rendered) using 22 MB" // app/Livewire/Debug/MemoryFootprintLogger.php:26

Test 1000000 
```
4. Keep refreshing page, to check for memory usage;
5. After some 20-30 refreshes check memory usage again:
```
"App\Livewire\Test (boot) using 71 MB" // app/Livewire/Debug/MemoryFootprintLogger.php:11

"App\Livewire\Test (mount) using 71 MB" // app/Livewire/Debug/MemoryFootprintLogger.php:16

"App\Livewire\Test (rendering) using 71 MB" // app/Livewire/Debug/MemoryFootprintLogger.php:21

"App\Livewire\Test (rendered) using 87 MB" // app/Livewire/Debug/MemoryFootprintLogger.php:26

Test 1000000 
```

It has grown significantly.

---

The same issue seems to happen if property is used instead of `#[Computed]` function.

Livewire template class
```php
    public array $largeArray = [];

    public function boot(): void
    {
        $this->largeArray = array_fill(0, 1000000, 'test');
    }

    public function render(): View
    {
        return view('livewire.test');
    }
```
Blade template
```blade
<div>
    Test {{ count($largeArray) }}
</div>

```

Initially usage of `22mb`, after multiple refreshes:
```
"App\Livewire\Test (boot) using 82 MB" // app/Livewire/Debug/MemoryFootprintLogger.php:11

"App\Livewire\Test (mount) using 82 MB" // app/Livewire/Debug/MemoryFootprintLogger.php:16

"App\Livewire\Test (rendering) using 82 MB" // app/Livewire/Debug/MemoryFootprintLogger.php:21

"App\Livewire\Test (rendered) using 82 MB" // app/Livewire/Debug/MemoryFootprintLogger.php:26

Test 1000000 
```

---

So is there something I'm doing fundamentally wrong? Or is this indeed an issue with Livewire? Because it seems what I'm doing here is pretty straight forward and would happen for all kinds of components in octane - it only becomes noticeable when I started working with larger amounts of data, but it seems the issue was there all along, for all of the components.

Could not reproduce this with plain blade templates, memory usage never accumulated.

As a temporary workaround, worker requests lifetime can be reduced, in hopes that it won't reach memory limit within that range.
By default worker is alive for `500` requests, as per [documentation](https://frankenphp.dev/docs/laravel/#laravel-octane).
When starting frankenphp this value can be adjusted. e.g,
```bash
php artisan octane:frankenphp --max-requests=20
```

**Comment by sts-ryan-holton** [comment]
I'm seeing a similar problem, copied your code into my component that's being loaded. On initial load:

> App\Livewire\Pages\Applications\ApplicationIndex (render) using 31 MB

But after around 10 requests:

> App\Livewire\Pages\Applications\ApplicationIndex (render) using 33 MB

Worth noting this is in my local environment via Laravel Sail, with just me, in reality, this problem could be so much worse.

**Reply by EdgarsJoja** [reply]
yep, I'm on Sail as well, but as you said - it's just one person and with larger amounts of data I would easily hit the "allowed memory exausted" error already. In production this could be a complete deal breaker for Livewire & Octane combo right now for me, if it indeed is a core issue.

**Comment by EdgarsJoja** [comment]
I have created simple repository demonstrating the issue.
FrankenPHP, Laravel, Livewire, Octane.
https://github.com/EdgarsJoja/livewire-octane-memory-leak-test

I don't know what has recently changed, but since upgrading to php 8.4, latest livewire, frankenphp, octane etc. the memory leak is even more noticable.

Initially the memory usage was somewhat fluctuating, with the overall trend increasing.
After upgrades, I can see that every single request in Livewire increases memory until crash, as demonstrated in this repository.

**Reply by EdgarsJoja** [reply]
@joshhanley  @calebporzio  sorry to tag you here, but perhaps you have some ideas with regards to this issue? I did try to find the source of the problem, but have come up with nothing so far. All the static variables, seem to be cleared between requests, singletons seem to be reset... Not sure where else to look.

**Reply by EdgarsJoja** [reply]
Hmm... may have actually found something this time. What seems to work is to manually unset properties & computed properties.
For example, putting this into component `boot` method:
```php
        defer(function () {
            // Computed properties have to be unset manually
           unset($this->myLargeComputedProperty);

            // This unsets regular properties
            foreach (get_object_vars($this) as $key => $val) {
                unset($this->$key);
            }
        });
```

This seems to free the memory for livewire components.

**Comment by joshhanley** [comment]
@EdgarsJoja @sts-ryan-holton thanks for reporting! I've submitted PR #10022 with a fix.
